File: src/main/java/com/pandus/leetcode/solutions/google/AddTwoNumbers.java
package com.pandus.leetcode.solutions.google;

public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode tail = dummy;
        int carry = 0;
        while (l1 != null || l2 != null || carry != 0) {
            int n1;
            if (l1 == null) {
                n1 = 0;
            } else {
                n1 = l1.val;
            }

            int n2;
            if (l2 == null) {
                n2 = 0;
            } else {
                n2 = l2.val;
            }

            int sum = n1 + n2 + carry;
            int newVal = sum % 10;
            carry = sum / 10;

            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
            tail.next = new ListNode(newVal);
            tail = tail.next;
        }
        return dummy.next;
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/top/interview/AddTwoNumbers.kt
package com.pandus.leetcode.solutions.top.interview

//Reference: https://leetcode.com/problems/add-two-numbers/
class AddTwoNumbers {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var tail: ListNode? = null
        var head: ListNode? = null

        var topNode = l1
        var bottomNode = l2

        var carry = 0 // 1 or 0
        while (topNode != null || bottomNode != null || carry != 0) {

            val top = topNode?.`val` ?: 0
            val bottom = bottomNode?.`val` ?: 0

            val sum = top + bottom + carry
            val newNum = sum % 10
            carry = if (sum > 9) 1 else 0

            val newNode = ListNode(newNum)
            if (tail == null) {
                tail = newNode
                head = tail
            } else {
                head!!.next = newNode
                head = newNode
            }

            topNode = topNode?.next
            bottomNode = bottomNode?.next
        }
        return tail
    }
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}